### PR TITLE
Adds support for basic Wasp `job` DSL parsing

### LINKED ISSUE #532: Parsing Wasp DSL for jobs (without cron)

### PR DESCRIPTION
**Pull request by shayneczyzewski**
# Description

This adds support for basic Wasp `job` DSL parsing, but provides no API implementation or end user functionality yet. That is covered by #533 

Closes #532

### CODE REVIEW

**Comment by shayneczyzewski** [comment]
That felt wayyyy too easy @Martinsos lol Am I missing something here, or did the magic you all did with TH injection just do its job very well? 🤔 :D

**Review by Martinsos — APPROVED** [review]
Ha awesome! As I said, it will be simple - TH does most of the heavy lifting for us :)!

The only thing I am wondering about is this `perform` name -> I believe you took it from RoR? I feel like I haven't seen it around much before so I feel a bit weird about it. But is it commonly used for jobs/workers? Are there better candidate? Smth like `execute`? or `do`? Or `fn`? Ha, none of those sounds better, does it? Ok no worries, let's keep `perform` then, I can't figure out anything better - if you have some idea I am all ears though!

**Comment by shayneczyzewski** [comment]
> Ha awesome! As I said, it will be simple - TH does most of the heavy lifting for us :)!
> 
> The only thing I am wondering about is this `perform` name -> I believe you took it from RoR? I feel like I haven't seen it around much before so I feel a bit weird about it. But is it commonly used for jobs/workers? Are there better candidate? Smth like `execute`? or `do`? Or `fn`? Ha, none of those sounds better, does it? Ok no worries, let's keep `perform` then, I can't figure out anything better - if you have some idea I am all ears though!

Good point to do a double check here. Indeed I took this from RoR, but I'll do a quick survey of names in this space and report back if something may be more common. 👍🏻 And yes, perhaps we just even use our own internal conventions like `fn` instead.

**Comment by Martinsos** [comment]
> > Ha awesome! As I said, it will be simple - TH does most of the heavy lifting for us :)!
> > The only thing I am wondering about is this `perform` name -> I believe you took it from RoR? I feel like I haven't seen it around much before so I feel a bit weird about it. But is it commonly used for jobs/workers? Are there better candidate? Smth like `execute`? or `do`? Or `fn`? Ha, none of those sounds better, does it? Ok no worries, let's keep `perform` then, I can't figure out anything better - if you have some idea I am all ears though!
> 
> Good point to do a double check here. Indeed I took this from RoR, but I'll do a quick survey of names in this space and report back if something may be more common. 👍🏻 And yes, perhaps we just even use our existing Wasp conventions like `fn` instead.

Although `fn` does kind of suck since it doesn't bring much info right? So that is why I didn't push on it. But sounds good, interested what you find out!

**Review by faassen — APPROVED** [review]

**Comment by shayneczyzewski** [comment]
> 

In the Ruby and Elixir worlds, `perform` is most dominant. In JS ones, you typically register via a `process` callback. Some others I saw that we can consider include `register`, `handler`, `do`, and `call`, but no real consensus on naming sadly. I'm good with whatever the group finds most intuitive.

File: waspc/test/AnalyzerTest.hs [hunk 1]
```diff
@@ -190,6 +196,18 @@ spec_Analyzer = do
...
+            [ ( "BackgroundJob",
+                Job.Job
+                  { Job.perform =
+                      ExtImport
+                        (ExtImportField "backgroundJob")
+                        (fromJust $ SP.parseRelFileP "jobs/baz.js")
+                  }
+              )
+            ]
+      takeDecls @Job <$> decls `shouldBe` Right expectedJob
```
Comment (sodic) [inline waspc/test/AnalyzerTest.hs:209]: I think you can safely remove `@Job` (same goes for most of the older tests).

File: waspc/src/Wasp/Analyzer/StdTypeDefinitions.hs [hunk 2]
```diff
@@ -41,5 +43,6 @@ stdTypes =
   TD.addDeclType @Route $
   TD.addDeclType @Query $
   TD.addDeclType @Action $
+  TD.addDeclType @Job $
```
Comment (matijaSos) [inline waspc/src/Wasp/Analyzer/StdTypeDefinitions.hs:46]: There is nothing to comment on in this file, I'm disappointed, it's too easy to add a new declaration :D TH is taking our jobs!

File: waspc/src/Wasp/AppSpec.hs [hunk 3]
```diff
@@ -73,6 +75,9 @@ getPages = getDecls @Page
 getRoutes :: AppSpec -> [(String, Route)]
 getRoutes = getDecls @Route
 
+getJobs :: AppSpec -> [(String, Job)]
+getJobs = getDecls @Job
```
[the remark this answers is not in the document]
Comment (matijaSos) [reply waspc/src/Wasp/AppSpec.hs:79]: nice catch!

[the remark this answers is not in the document]
Comment (shayneczyzewski) [reply waspc/src/Wasp/AppSpec.hs:79]: Ah cool thanks Filip! I can try removing and see!

File: waspc/test/AnalyzerTest.hs (hunk 1 above)
Comment (shayneczyzewski) [reply waspc/test/AnalyzerTest.hs:209]: Thanks, will give it a try!

**Review by matijaSos — APPROVED** [review]
wow! As I also mentioned in the comments, I am shocked how easy this was (to add a new declaration) and sad there was nothing for me to comment on :D JK, this looks really cool!

File: waspc/test/AnalyzerTest.hs (hunk 1 above)
Comment (shayneczyzewski) [reply waspc/test/AnalyzerTest.hs]: I was able to remove all type applications except this one, which the compiler could not automatically infer:

```haskell
takeDecls @Route <$> analyze source
  `shouldBe` Left (TypeError $ TC.mkTypeError (ctx (1, 34) (1, 48)) $ TC.UndefinedIdentifier "NonExistentPage")
```

Thanks! 👍🏻

**Pull request merged**